File: backend/app/services/chat_service.py
from sqlalchemy.orm import Session as DBSession
from typing import List, Optional
from fastapi import HTTPException, status

from ..models.chat_model import Chat, MessageRole
from ..models.session_model import Session

class ChatService:
    
    @staticmethod
    def create_chat(db: DBSession, session_id: str, role: MessageRole, content: str) -> dict:
        # Verifica se a session exists
        session = db.query(Session).filter(Session.id == session_id).first()
        if not session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Session not found"
            )
        
        chat = Chat(
            session_id=session_id,
            role=role,
            content=content
        )
        db.add(chat)
        db.commit()
        db.refresh(chat)
        
        return {
            "id": chat.id,
            "session_id": chat.session_id,
            "role": chat.role.value,
            "content": chat.content,
            "created_at": chat.created_at
        }
    
    @staticmethod
    def get_chats_by_session(db: DBSession, session_id: str, skip: int = 0, limit: int = 100) -> List[dict]:
        # Verifica se a session exists
        session = db.query(Session).filter(Session.id == session_id).first()
        if not session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Session not found"
            )
        
        chats = db.query(Chat).filter(Chat.session_id == session_id).offset(skip).limit(limit).all()
        
        return [
            {
                "id": chat.id,
                "session_id": chat.session_id,
                "role": chat.role.value,
                "content": chat.content,
                "created_at": chat.created_at
            }
            for chat in chats
        ]
    
    @staticmethod
    def get_chat_by_id(db: DBSession, chat_id: str) -> dict:
        chat = db.query(Chat).filter(Chat.id == chat_id).first()
        
        if not chat:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Chat not found"
            )
        
        return {
            "id": chat.id,
            "session_id": chat.session_id,
            "role": chat.role.value,
            "content": chat.content,
            "created_at": chat.created_at
        }
    
    @staticmethod
    def update_chat(db: DBSession, chat_id: str, content: Optional[str] = None) -> dict:
        chat = db.query(Chat).filter(Chat.id == chat_id).first()
        
        if not chat:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Chat not found"
            )
        
        if content is not None:
            chat.content = content
        
        db.commit()
        db.refresh(chat)
        
        return {
            "id": chat.id,
            "session_id": chat.session_id,
            "role": chat.role.value,
            "content": chat.content,
            "created_at": chat.created_at
        }
    
    @staticmethod
    def delete_chat(db: DBSession, chat_id: str) -> None:
        chat = db.query(Chat).filter(Chat.id == chat_id).first()
        
        if not chat:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Chat not found"
            )
        
        db.delete(chat)
        db.commit()

File: frontend/src/services/chatService.ts
import { Chat } from "@/models/chat";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:8000/api/v1';

interface CreateChatRequest {
  session_id: string;
  question: string;
  k?: number;
}

export function ChatService() {
  const createChat = async (request: CreateChatRequest) => {
    const { session_id, question, k = 50 } = request;
    const response = await fetch(`${API_BASE_URL}/chats?session_id=${session_id}&question=${encodeURIComponent(question)}&k=${k}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
    });
    return response.json();
  };

  const getChatsBySession = async (
    sessionId: string, 
    skip: number = 0, 
    limit: number = 100
  ): Promise<Chat[]> => {
    const response = await fetch(`${API_BASE_URL}/chats/session/${sessionId}?skip=${skip}&limit=${limit}`);
    return response.json();
  };

  const getChat = async (chatId: string): Promise<Chat> => {
    const response = await fetch(`${API_BASE_URL}/chats/${chatId}`);
    return response.json();
  };

  const updateChat = async (chatId: string, content?: string): Promise<Chat> => {
    const response = await fetch(`${API_BASE_URL}/chats/${chatId}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    });
    return response.json();
  };

  const deleteChat = async (chatId: string): Promise<void> => {
    await fetch(`${API_BASE_URL}/chats/${chatId}`, { method: 'DELETE' });
  };

  return { createChat, getChatsBySession, getChat, updateChat, deleteChat };
}